### PR TITLE
Block Support Panels - Make block support tools panels compatible with multi-selection

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: 
- https://github.com/WordPress/gutenberg/issues/37188

Depends on: 
- https://github.com/WordPress/gutenberg/pull/37273

## Description

The recent switch to [using grouped inspector controls for the typography](https://github.com/WordPress/gutenberg/pull/33744/files#diff-068538657061c47f0a58cfc331259bbeabe11bcc0c576a0cc496d262e5b0038cR132-R136) block support panel meant the typography controls were no longer included within the normal `InspectorControls` slot. As such, they didn't display when multi-block selections were made.

This PR adds the new grouped inspector controls slots into the block inspector for multi-selections. It also updates the `BlockSupportToolsPanel` to be compatible with multiple selected blocks. 

## How has this been tested?

1. Create a new post and add multiple paragraph blocks
2. Select a single paragraph block and adjust its typography ensuring that still functions
3. Now select multiple paragraph blocks together
4. The typography controls should still be present in the sidebar (they will display the options as per the first block selected in order of position on the page)
5. Adjust the typography settings and ensure all selected blocks update accordingly
6. Save the post and check blocks on the frontend
7. Test resetting individual controls as well as the "reset all" option.

## Screenshots 

MultiSelectTypography

## Types of changes
Bug fix.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
I've split out the `ToolsPanel` specific changes from this PR into https://github.com/WordPress/gutenberg/pull/37273.

**Comment by ramonjd** [comment]
Works as described for me:

✅ Select a single paragraph block and adjust its typography ensuring that still functions
✅ Now select multiple paragraph blocks together
✅ The typography controls should still be present in the sidebar (they will display the options as per the first block selected in order of position on the page)
✅ Adjust the typography settings and ensure all selected blocks update accordingly
✅ Save the post and check blocks on the frontend
✅ Test resetting individual controls as well as the "reset all" option.

I also tested multi-selecting other text-blocks (heading etc): the typography panel doesn't appear.

> they will display the options as per the first block selected in order of position on the page

At first it threw me to see subsequent paragraphs' typography settings switch to the initial block's 

Kapture 2021-12-13 at 09 12 38

I think my brain expected to see only the changed settings take effect, for example, if I only altered line height, then only the line height would change.

But I understand there needs to be some base value in the controls when multi-selecting and using the initial block's seems to me to be the most logical.

I'm not arguing for a change in functionality, just documenting. :)

File: packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js
Comment (ramonjd) [inline packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: Could you explain this comment to me? 

Does it mean `selectedBlockClientId` is `null` when `hasMultiSelection() === true` and ... 

or should that "and" be an "is"?

Comment (ramonjd) [reply packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: Update: I probably should have started looking at https://github.com/WordPress/gutenberg/pull/37273 before this one 😅

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: Apologies for the confusion. 

I've pushed a commit with an updated comment hopefully clarifying how this works.

The PR was split to isolate the `ToolsPanel` related changes as has been requested in the past.

**Review by andrewserong — COMMENTED** [review]
Nice work @aaronrobertshaw! The overall approach looks good to me, and the ToolsPanel / ToolsPanelItem behaviour worked correctly for me.

I noticed one issue with keyboard handling, which is that when updating a UnitControl with multiple blocks selected, pressing `Backspace` or `Delete` on my keyboard within the field, unintentionally winds up deleting the blocks instead of just the character in the input field:

Kapture 2021-12-13 at 11 17 19

I wasn't _too_ sure where the fix should go for this, whether the individual UnitControl component needs to add an `event.stopPropagation`, or if we should do that within the BlockSupportToolsPanel?

**Comment by aaronrobertshaw** [comment]
Thanks for testing @ramonjd and @andrewserong 🙇 

> I also tested multi-selecting other text-blocks (heading etc): the typography panel doesn't appear.

By this, I take it you mean mixing block types didn't display the typography panel? If I select multiple heading blocks I still get the typography panel in the block inspector.

> I noticed one issue with keyboard handling, which is that when updating a UnitControl with multiple blocks selected, pressing Backspace or Delete on my keyboard within the field, unintentionally winds up deleting the blocks instead of just the character in the input field:

Nice catch! 

> I wasn't too sure where the fix should go for this, whether the individual UnitControl component needs to add an event.stopPropagation, or if we should do that within the BlockSupportToolsPanel?

This appears to affect more than just the `UnitControl`. Any field accepting keyboard input suffers from the same problem e.g. `TextControl`, `ColorPicker` inputs etc. There is a block editor keyboard shortcut for the `del` and `backspace` keys while more than one block is selected. It is this, that is being triggered and causing the block deletion.

I don't think it is feasible to rely on any input within the `BlockInspector` preventing propagation of keyboard events for just the `del` or `backspace` keys. Especially once we take into account 3rd party blocks etc. If all key presses weren't propagated we'd also then block other keyboard shortcuts such as `cmd + s` for saving etc.

This issue looks to be unrelated to this PR and probably needs a broader solution. It exists on trunk with the `ColorPicker`. Click on the color swatch, toggle the edit button for the color value in the picker and press the `del` key there to replicate. 

There was a [PR](https://github.com/WordPress/gutenberg/pull/34539) a few months ago switching shortcut handling to React events but this issue appears to pre-date that. I went back several months without finding a point that you could use `del` or `backspace` in a block inspector field without triggering the deletion of the multi-selection.

I've created an issue for this here: https://github.com/WordPress/gutenberg/issues/37316

Moving forward, I don't think this issue needs to block this PR or https://github.com/WordPress/gutenberg/pull/37273. What do you think?

@youknowriad do you have any thoughts on how best to prevent only the [`core/block-editor/delete-multi-selection`](https://github.com/WordPress/gutenberg/blob/d03b7568654b5f3a5972a18c8ccb9503d4ed5a4f/packages/block-editor/src/components/block-tools/index.js#L96) event when attempting to clear inputs via the keyboard?

**Comment by ramonjd** [comment]
> By this, I take it you mean mixing block types didn't display the typography panel? If I select multiple heading blocks I still get the typography panel in the block inspector.

Yes, thanks for interpreting my vague comment. 😆 I can confirm that selecting multiple text blocks of the same type (e.g, Heading) works as advertised as well.

**Comment by ramonjd** [comment]
I think I might have found another peculiarity, and it would be good to get a confidence check.

Selecting blocks by dragging upwards appears to reset styles.

It appears to happen when the last empty paragraph block is included in the multiple select (and thereafter even if the empty paragraph block isn't included in the selection) , and only when dragging upwards and not downwards.

Notice that the Typography panel is empty.

Kapture 2021-12-13 at 22 37 12

I checked on trunk and I can't replicate using color controls.

Kapture 2021-12-13 at 22 25 42

**Comment by aaronrobertshaw** [comment]
> I think I might have found another peculiarity, and it would be good to get a confidence check.

Thanks for bringing that one to light @ramonjd 👍 

I might have hit that one as well early on when I was putting this together but when I got to testing I couldn't replicate it. 

I'll dig into this tomorrow. Given it isn't happening with the color panel it appears related to the `ToolsPanelItem` registrations. Also, I'll hold off on merging https://github.com/WordPress/gutenberg/pull/37273 until I work out what is going on.

**Comment by mtias** [comment]
Marking as backport considering it's a regression.

**Comment by aaronrobertshaw** [comment]
> It appears to happen when the last empty paragraph block is included in the multiple select (and thereafter even if the empty paragraph block isn't included in the selection) , and only when dragging upwards and not downwards.

After digging into this, I found the issue arises once more than 2 blocks are selected in reverse order. I believe it caused panel items to deregister unexpectedly along with triggering the item's `onDeselect` callback which is what resulted in the reset attributes.

I've updated the logic around panel item registration and deregistration in https://github.com/WordPress/gutenberg/pull/37273. I think this has solved this particular issue as it works for me in my testing so far. Hopefully, you get the same results 🤞.

> I think my brain expected to see only the changed settings take effect, for example, if I only altered line height, then only the line height would change.

I felt the same @ramonjd. This might be something we can iterate on as a follow-up to this?

**Comment by ramonjd** [comment]
> I think this has solved this particular issue as it works for me in my testing so far. Hopefully, you get the same results 🤞.

Nice, thanks a lot! I can no longer reproduce the reverse selection bug  🍺 



Kapture 2021-12-17 at 10 01 19




> I felt the same @ramonjd. This might be something we can iterate on as a follow-up to this?

Definitely 👍

File: packages/block-editor/src/components/block-inspector/index.js
Comment (ramonjd) [inline packages/block-editor/src/components/block-inspector/index.js:74]: Is here a good opportunity to flip Dimensions and Border panels according to the order defined in https://github.com/WordPress/gutenberg/issues/35541? (Dimensions before Border)

**Review by ramonjd — APPROVED** [review]
Took this for another spin:

✅ Select a single paragraph block and adjust its typography ensuring that still functions
✅ Now select multiple paragraph blocks together
✅ The typography controls should still be present in the sidebar (they will display the options as per the first block selected in order of position on the page)
✅ Adjust the typography settings and ensure all selected blocks update accordingly
✅ Save the post and check blocks on the frontend
✅ Test resetting individual controls as well as the "reset all" option.
✅ Multiselecting text blocks, upwards and downwards, has no effect on the above functionality.


LGTM

File: packages/block-editor/src/components/block-inspector/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:74]: Absolutely. Done ✅

**Review by andrewserong — APPROVED** [review]
Thanks @aaronrobertshaw, I just re-tested this and it's working nicely for me: 

✅ Adjusting a single paragraph block in isolation works
✅ Adjusting multiple paragraph blocks together works with settings being applied to the selection, and resetting individual controls and "reset all" works as to be expected

The only issue I ran into is the issue of keyboard shortcuts firing from the sidebar which we're looking into separately in #37326, so I don't think that's a blocker for getting this PR in.

LGTM! 🎉

**Comment by aaronrobertshaw** [comment]
> I felt the same @ramonjd. This might be something we can iterate on as a follow-up to this?

I've looked into the way this worked prior to the `ToolsPanel` switch for the typography panel. After this PR is merged, we are back to the same functionality. I'm not sure we should change that behaviour if it is what people have come to expect.

Of course, if it is proven desirable to refine it, I'll happily revisit this.

<details>
<summary>Demo of functionality before Typography ToolsPanel switch</summary>

https://user-images.githubusercontent.com/60436221/146488702-f571fa80-91c3-46b4-a2c9-944cedd7de3b.mp4

</details>

**Pull request merged**
